jetpack/src/connection.rs

```diff
@@ -0,0 +1,354 @@
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use wp_api::{
+    request::RequestExecutor, users::UserId, ParsedUrl, WpApiError, WpAuthentication, WpErrorCode,
+};
+use wp_com::{
+    client::WpComApiClient, jetpack_connection::JetpackRemoteConnectionParams, WpComSiteId,
+};
+
+use crate::client::JetpackApiClient;
+
+#[derive(Debug, Serialize, uniffi::Record)]
+pub struct JetpackConnectionParams {
+    /// origination of the request, e.g. "jetpack-app"
+    pub from: String,
+    /// the plugin being connected, e.g. "jetpack"
+    pub plugin_slug: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+#[serde(rename_all = "camelCase")]
+pub struct JetpackConnectionRegisterResult {
+    pub authorize_url: String,
+}
+
+impl JetpackConnectionRegisterResult {
+    fn blog_id(&self) -> Option<WpComSiteId> {
+        self.authorize_url
+            .parse::<url::Url>()
+            .ok()
+            .and_then(|url| {
+                url.query_pairs()
+                    .find(|(key, _)| key == "client_id")
+                    .map(|(_, value)| value.to_string())
+            })
+            .and_then(|str| str.parse().ok())
+    }
+}
+
+#[derive(Debug, Serialize, uniffi::Record)]
+pub struct JetpackRemoteProvisionParams;
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+pub struct JetpackRemoteProvisionResult {
+    pub jp_version: String,
+    pub redirect_uri: String,
+    pub user_id: UserId,
+    pub user_email: String,
+    pub user_login: String,
+    pub scope: String,
+    pub secret: String,
+    pub is_active: bool,
+}
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+#[serde(rename_all = "camelCase")]
+pub struct JetpackConnection {
+    pub is_active: bool,
+    pub is_user_connected: bool,
+}
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+#[serde(rename_all = "camelCase")]
+pub struct JetpackConnectionData {
+    pub current_user: JetpackConnectionUser,
+    pub connection_owner: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+#[serde(rename_all = "camelCase")]
+pub struct JetpackConnectionUser {
+    pub id: UserId,
+    pub username: String,
+    pub is_connected: bool,
+    pub is_master: bool,
+    pub blog_id: Option<WpComSiteId>,
+}
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+pub struct JetpackConnectionCheck {
+    pub status: String,
+}
+
+/// Connect a WordPress site to Jetpack.
+///
+/// Please note, the endpoints used are available on Jetpack 14.2 and above.
+#[derive(Debug, uniffi::Object)]
+pub struct JetpackConnectionClient {
+    request_executor: Arc<dyn RequestExecutor>,
+    jetpack_client: JetpackApiClient,
```

Comment:
This setup seems a little backwards to me. Shouldn't we add `JetpackConnectionClient` to `JetpackApiClient` instead? Instead of `JetpackApiClient`, this type should probably take a `ConnectionRequestExecutor` and we should remove the `connection: Arc<ConnectionRequestExecutor>` field from `JetpackApiClient`.

All these suggestions are under the assumption that we are not supposed to use the `ConnectionRequestExecutor` directly, but rather use the `JetpackConnectionClient` because it provides further processing. Let me know if that's a misunderstanding.

Reply:
I could see either approach making sense. I always thought `JetpackApiClient` and `WpApiClient` would be the overarching container, but it can also be treated as the clients that allow raw requests as you've done so here.

I am fine with leaving it as is for now and we can re-consider it if we find any other pros/cons to either approach.